HandEyeCalibration/Transform/Transform.cpp

```diff
@@ -0,0 +1,330 @@
+        std::cout << "\nConverting transformation matrix to rotation matrix and translation vector" << std::endl;
+        const Eigen::Matrix3d rotationMatrix = TransformUtils::cv2eigen(poseState(cv::Range(0, 3), cv::Range(0, 3)));
+        const Eigen::Vector3d translationVector = TransformUtils::cv2eigen(poseState(cv::Range(0, 3), cv::Range(3, 4)));
+        std::cout << "\nRotationMatrix:\n" << rotationMatrix << std::endl;
+        fileStorage.write("RotationMatrix", TransformUtils::eigen2cv(rotationMatrix));
+        std::cout << "\nTranslationVector:\n" << translationVector << std::endl;
+        fileStorage.write("TranslationVector", TransformUtils::eigen2cv(translationVector));
+
+        // Rotation Matrix in Transformation Matrix to Axis-angle
+        std::cout << "\nConverting rotation matrix to axis-angle" << std::endl;
+        Eigen::AngleAxisd axisAngle = Eigen::AngleAxisd(rotationMatrix);
+        Eigen::Vector4d axisAngleReordered = TransformUtils::reorderAxisAngle(axisAngle);
+        std::cout << "\nAxis-angle:\n" << axisAngleReordered << std::endl;
+        fileStorage << "Axis-Angle format"
+                    << "ux uy uz theta";
+        fileStorage.write("AxisAngle", TransformUtils::eigen2cv(axisAngleReordered));
+
+        // Axis-angle to Rotation Vector
+        std::cout << "\nConverting axis-angle to rotation vector" << std::endl;
+        if(!axisAngle.axis().isUnitary())
+        {
+            std::cout << "Note! Axis in axis-angle pair is not unitary" << std::endl;
+        }
+        const auto rotationVector = axisAngle.angle() * axisAngle.axis();
+        fileStorage << "Rotation vector format"
+                    << "rx ry rz";
+        std::cout << "\nRotationVector:\n" << rotationVector << std::endl;
+        fileStorage.write("RotationVector", TransformUtils::eigen2cv(rotationVector));
+
+        // Rotation Matrix to Quaternion
+        std::cout << "\nConverting rotation matrix to quaternion" << std::endl;
+        Eigen::Quaterniond quaternion = Eigen::Quaterniond(rotationMatrix);
+        Eigen::Vector4d quaternionReordered = quaternion.coeffs();
+        std::cout << "\nQuaternion:\n" << quaternionReordered << std::endl;
+        fileStorage << "Quaternion format"
+                    << "qx qy qz qw ";
+        fileStorage.write("Quaternion", TransformUtils::eigen2cv(quaternionReordered));
+
+        // Rotation Matrix to Roll-Pitch-Yaw
+        const TransformUtils::Rotation_s rotations[4] = {
+            { TransformUtils::Rotation::ZYX_INTRINSIC, "ZYX_INTRINSIC", "RollPitchYawIntrinsicZYX" },
+            { TransformUtils::Rotation::XYZ_EXTRINSIC, "XYZ_EXTRINSIC", "RollPitchYawExtrinsicXYZ" },
+            { TransformUtils::Rotation::XYZ_INTRINSIC, "XYZ_INTRINSIC", "RollPitchYawIntrinsicXYZ" },
+            { TransformUtils::Rotation::ZYX_EXTRINSIC, "ZYX_EXTRINSIC", "RollPitchYawExtrinsicZYX" }
+        };
+        Eigen::Array3d rollPitchYaw[4];
+        for(uint32_t i = 0; i < 4; i++)
```

Comment:
The embedded programmer in me would have wanted to define a uint8_t :)... I need to learn that saving a few bits probably isn't worth it?

Reply:
I've really tried to figure out if I can use `std::transform`, but I believe there are several problems. I would have to create a new function to use in `std::transform`, because I want to keep the `std::cout` prints, and I want to save the output of each individual conversion to a file. This means I would have to pass in `fileStorageOut` and I'm not sure how to pass an argument to the function used in `std::transform`, that's not part of the array.

Reply:
For `std::transform` you typically use a lambda. Then you can capture whatever locals you need. I can show you irl if you want.

By the way we should think about the data types for `rotations`, `rollPitchYaw` etc., raw arrays might not be the best choice. Let's discuss that irl too if you're in the office later today.